stores/metadata.go

```diff
@@ -877,72 +887,60 @@ WHERE c.fcid = ?
 }
 
 func (s *SQLStore) SetContractSet(ctx context.Context, name string, contractIds []types.FileContractID) error {
-	fcids := make([]fileContractID, len(contractIds))
-	for i, fcid := range contractIds {
-		fcids[i] = fileContractID(fcid)
+	wanted := make(map[fileContractID]struct{})
+	for _, fcid := range contractIds {
+		wanted[fileContractID(fcid)] = struct{}{}
 	}
 
 	var diff []fileContractID
 	err := s.retryTransaction(func(tx *gorm.DB) error {
-		// fetch current contracts
-		var dbCurrentContracts []fileContractID
+		// fetch contract set
+		var cs dbContractSet
 		err := tx.
-			Model(&dbContract{}).
-			Select("contracts.fcid").
-			Joins("LEFT JOIN contract_set_contracts csc ON csc.db_contract_id = contracts.id").
-			Joins("LEFT JOIN contract_sets cs ON cs.id = csc.db_contract_set_id AND cs.name = ?", name).
-			Group("contracts.fcid").
-			Scan(&dbCurrentContracts).
+			Where(dbContractSet{Name: name}).
+			Preload("Contracts").
```

Comment:
Could you skip the `Preload` here and then just fetch all contracts with an ID in `contractIds`?
Then you might be able to do `tx.Model(&cs).Association("Contracts").Replace(&contracts)` instead of `Delete` and `Append`.
Might also be able to skip the `Unscoped`. Afaik it's only useful when you have soft deletion.

Reply:
I rewrote it to try hooks and then afterwards kept it like this because I like that collecting the `diff` is very straightforward if we do it like this. The `unscoped` is necessary to have `Delete` effectively get rid of the join table entity. I don't really mind reverting to what you had originally, do you prefer the old way? Only need to change the `LEFT JOIN` to `INNER JOIN`s.

Reply:
I'm not saying go back to the old one but change it to something like this

```go
// fetch contract set
var cs dbContractSet
err := tx.
	Where(dbContractSet{Name: name}).
	FirstOrCreate(&cs).
	Error
if err != nil {
	return err
}

// fetch contracts
var contracts []dbContract
err : tx.
      Where("fcid", wanted).
      Find(&contracts).
      Error
if err != nil {
      return err
}

// update the association
if err := tx.Model(&cs).Association("Contracts").Unscoped().Replace(&contracts); err != nil {
	return err
}
```

Reply:
I got rid of the unscoped and add/delete but the `Preload` is necessary 🤷‍♂️

Reply:
Ahhh I see. I thought you'd also get rid of the invalidation and as a result of the `Preload`. Since we got timestamps now.